src/ripple/app/misc/tests/AmendmentTable.test.cpp

```diff
@@ -375,8 +377,401 @@ class AmendmentTable_test final : public beast::unit_test::suite
+        auto const amendmentNames (
+            populateTable (*table, m_knownAmendmentPairs));
+
+        auto const validators = makeValidators (10);
+
+        std::vector <std::pair <uint256, int>> votes;
+        std::vector <uint256> ourVotes;
+        enabledAmendments_t enabled;
+        majorityAmendments_t majority;
+
+        // Week 1: We should vote for all known amendments not enabled
+        doRound (*table, 1,
+            validators,
+            votes,
+            ourVotes,
+            enabled,
+            majority);
+        expect (ourVotes.size() == amendmentNames.size(), "Did not vote");
+        expect (enabled.empty(), "Enabled amendment for no reason");
+        for (auto const& i : amendmentNames)
+            expect(majority.find(i.id()) == majority.end(), "majority detected for no reaosn");
+
+        // Now, everyone votes for this feature
+        for (auto const& i : amendmentNames)
+            votes.emplace_back (i.id(), 256);
+
+        // Week 2: We should recognize a majority
+        doRound (*table, 2,
+                validators,
+                votes,
+                ourVotes,
+                enabled,
+                majority);
+        expect (ourVotes.size() == amendmentNames.size(), "Did not vote");
+        expect (enabled.empty(), "Enabled amendment for no reason");
+        for (auto const& i : amendmentNames)
+            expect (majority[i.id()] == weekTime(2), "majority not detected");
+
+        // Week 5: We should enable the amendment
+        doRound (*table, 5,
+                validators,
+                votes,
+                ourVotes,
+                enabled,
+                majority);
+        expect (enabled.size() == amendmentNames.size(), "Did not enable");
+
+        // Week 6: We should remove it from our votes and from having a majority
+        doRound (*table, 6,
+                validators,
+                votes,
+                ourVotes,
+                enabled,
+                majority);
+        expect (enabled.size() == amendmentNames.size(), "Disabled");
+        expect (ourVotes.empty(), "Voted after enabling");
+        for (auto const& i : amendmentNames)
+            expect(majority.find(i.id()) == majority.end(), "majority not removed");
+    }
+
+    // Detect majority at 80%, enable later
+    void testDetectMajority ()
+    {
+        testcase ("detectMajority");
+        auto table (makeTable (2));
+
+        uint256 testAmendment;
+        testAmendment.SetHex("6dcd528f057711c5d26b57be28e23dfa");
+        table->addKnown({testAmendment, "testAmendment"});
+
+        auto const validators = makeValidators (16);
+
+        enabledAmendments_t enabled;
+        majorityAmendments_t majority;
+
+        for (int i = 0; i <= 17; ++i)
+        {
+            std::vector <std::pair <uint256, int>> votes;
+            std::vector <uint256> ourVotes;
+
+            if ((i > 0) && (i < 17))
+                votes.emplace_back (testAmendment, i * 16);
```

Comment:
@JoelKatz I don't understand the use of votes here. It looks like votes is per validator, but I think it should be per amendment. Looking at `AmendmentTable.test.cpp` line 423, doesn't this cause a validator to try to vote for the same amendment multiple times? Am I reading the code correctly?